File: butterbee/src/butterbee/port_finder.js
import { Ok, Error } from "./gleam.mjs";
import * as net from "net";

export function new_port() {
	return new Promise((resolve) => {
		const server = net.createServer();

		server.listen(0, "127.0.0.1", () => {
			const port = server.address().port;
			server.close(() => {
				resolve(new Ok(port));
			});
		});

		server.on("error", (err) => {
			resolve(new Error(new BindError(err.message)));
		});
	});
}

File: butterbee/src/butterbee/port_finder.erl
-module(port_finder).

-export([new_port/0]).

new_port() ->
    case gen_tcp:listen(0, [{ip, {127, 0, 0, 1}}]) of
        {ok, ListenSocket} ->
            case inet:sockname(ListenSocket) of
                {ok, {_, Port}} ->
                    gen_tcp:close(ListenSocket),
                    {ok, Port};
                {error, Reason} ->
                    gen_tcp:close(ListenSocket),
                    {error, {bind_error, atom_to_binary(Reason)}}
            end;
        {error, Reason} ->
            {error, {socket_error, atom_to_binary(Reason)}}
    end.
